Valid moves

jamesfoster85 opened the pull request:
- Created an accessible_squares method on the Piece model, which returns the squares that are 1) on the board and 2) either unoccupied or contain an opponents piece.  Legal_moves for each piece type can call this instead of ALL_SQUARES.

- When testing rook moves, I commented out the part of the is_obstructed? method that raises an exception, because it was resulting in Errors in the tests.  It didn't seem like we needed an exception, because all cases that could result in an exception would be invalid moves for other reasons and would be caught by other methods.

- legal_move? method now functions appropriately for both king and rook.

lasley commented on app/models/bishop.rb:
```diff
@@ -1,2 +1,7 @@
 class Bishop < Piece
+	def legal_moves
+		accessible_squares.reject{|s| is_obstructed?(s[0],s[1])}.select do |s|
```
Is there a way to centralize this repeated block?

lasley commented on test/models/knight_test.rb:
```diff
@@ -0,0 +1,64 @@
+  test 'Illegal move for all piece types' do
+    assert_not @knight.legal_move?(7, 5)
+  end
+
```
On board?

lasley commented on test/models/bishop_test.rb:
```diff
@@ -0,0 +1,40 @@
+  test 'Illegal move, own piece occupies square' do
+    assert_not @bishop.legal_move?(6, 1)
+  end
+
```
On board?

lasley commented on test/models/king_test.rb:
```diff
@@ -4,21 +4,21 @@ class KingTest < ActiveSupport::TestCase
     @game = Game.create(:name => 'lolomg', :white_player_id => 1)
-    @king = @game.pieces.find_by(type: 'King', player_id: 1)
+    @king = @game.pieces.create(type: 'King', x_position: 4, y_position: 2)
   end
```
On board?

lasley commented on test/models/queen_test.rb:
```diff
@@ -0,0 +1,44 @@
+  test 'Illegal move, own piece occupies square' do
+    assert_not @queen.legal_move?(3, 1)
+  end
+
```
What about `(5, 4)` (illegal diagonal)?

lasley commented on test/models/bishop_test.rb:
```diff
@@ -0,0 +1,40 @@
+  test 'Legal diagonal move' do
+    assert @bishop.legal_move?(4, 5)
+  end
+
```
What about `(4, 4)` (illegal diagonal)?

jamesfoster85 commented:
Added all of those tests (good catches, all of them), moved the repeated code into a central method on piece.rb.  I've started working on pawns but I haven't finished the method so I commented out the failing tests for now.  Once we have a working method we'll need to remember to uncomment and run them.

jamesfoster85 commented:
Finished the pawn logic.  Note:  Because a pawn cannot ...
- move backwards
- move sideways
- move diagonally unless capturing
- occupy the last rank of a file (because it is immediately promoted and thus no longer a pawn)
... I did not include a test for 'on the board' because any move that would be off the board should already be invalid for other reasons.

lasley commented:
:+1:

kenwmak commented on test/models/king_test.rb:
```diff
@@ -1,24 +1,33 @@
-    @game = Game.create(:name => 'lolomg', :white_player_id => 1)
-    @king = @game.pieces.find_by(type: 'King', player_id: 1)
+    @game = Game.create(name: 'lolomg', white_player_id: 1)
+    @king = @game.pieces.create(type: 'King', x_position: 4, y_position: 2)
```
James, how come you changed using ```create``` instead of ```find_by``` for ```@king```? I'm trying to resolve merge conflict now with my changes.

jamesfoster85 replied:
Ah.  I just needed a king in the middle of the board to test all the moves since the king in the starting position is blocked on all sides.  We could do
```
def game_and_kings
  @game = Game.create(name: 'lolomg', white_player_id: 1)
  @moves_king = @game.pieces.create(type: 'King', x_position: 4, y_position: 2)
  @castle_king = @game.pieces.find_by(type: 'King', player_id: 1)
end
```
And use the moves_king in the tests that I wrote and the castle_king in the tests you wrote.

kenwmak replied:
Alright, I'll do that.

lasley replied:
:+1:

jamesfoster85 commented:
Ah.  I just needed a king in the middle of the board to test all the moves since the king in the starting position is blocked on all sides.  We could do
```
def game_and_kings
  @game = Game.create(name: 'lolomg', white_player_id: 1)
  @moves_king = @game.pieces.create(type: 'King', x_position: 4, y_position: 2)
  @castle_king = @game.pieces.find_by(type: 'King', player_id: 1)
end
```
And use the moves_king in the tests that I wrote and the castle_king in the tests you wrote.
